### PR TITLE
feat(helm): Add --reset-then-reuse-values flag to 'helm upgrade'

### LINKED ISSUE #3957: Upgrade to new chart version and --set a value at the same time with --reuse-values ignores changes from values.yaml
Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:21:50Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.2", GitCommit:"bdaeafa71f6c7c04636251031f93464384d54963", GitTreeState:"clean", BuildDate:"2017-10-24T19:38:10Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
Kubeadm cluster on vsphere

I can demonstrate the problem with the chart created with `helm create`, but it is not specific to that chart in any way

```bash
helm create temp
cd temp
helm upgrade temp . --install --set replicaCount=2
# at this point I have a release with custom values I want to preserve
#
# but let's say a new version of the chart becomes available that adds new defaults to values.yaml
echo "hello: world" >> values.yaml
echo "goodbye: world" >> values.yaml

# I want to use this new version, but the defaults for one of the new values doesn't work for me
# so I can try to upgrade
helm upgrade temp . --set goodbye=nobody --install --debug --dry-run
# whoops, I would lose my custom settings if I ran that
# let's reuse values (that's what it's for, right?)
helm upgrade temp . --install --set goodbye=nobody --reuse-values  --debug --dry-run
# now replicaCount is still set, but, weirdly, hello:world is missing from the computed values
# seems like --reuse-values doesn't use values introduced by the new version of the chart
#
# if I didn't need to --set any values, I could run
helm upgrade temp . --install --debug --dry-run
# replicaCount is still 2, but the setting I wanted to set is not set, forcing me to upgrade in two steps
```

This behavior is unexpected, and will be hard to people to debug if they are not very familiar with all the new values introduced by a new chart version.

### LINKED ISSUE #8085: Feature request: --reset-then-reuse flag for upgrade
Currently Helm implements two ways of handling values when updating an existing release: reusing the existing user-supplied values if new ones are not supplied, or not reusing them if new ones are supplied.  However, the way these are implemented and the way `--reuse-values` and `--reset-values` change that default behavior is sometimes confusing; see, for example:

https://medium.com/@kcatstack/understand-helm-upgrade-flags-reset-values-reuse-values-6e58ac8f127e

(And see one of the commenters pointing out that even after this article-length treatment of the subject, the writer missed an important subtlety!)

What I think would be nice for users is a flag that means "take the existing chart's values, merge these new values I am giving you explicitly, and do not update values from the chart unless they are new to the chart since the version stored in the target release".  (Alternately one can think of this as "`--reset-values` followed by `--reuse-values` followed by user-supplied `--set*`values")

This seems like it would be less confusing in the majority of cases: if I give a value, it is updated; if I don't give it and it exists in the release, it is reused; if I don't give it and it doesn't exist in the release, it is set from the updated chart.  It would eliminate the case of "I did not specify this value as part of my upgrade but it changed anyway" which appears to be where much of the confusion strikes.

### PR DESCRIPTION
**Pull request by okhoshi**
**What this PR does / why we need it**:
When `--reset-then-reuse-values` is used on 'helm upgrade', the chart's values will be reset to the values of the deployed chart while the current release's values will be reused and merged with the values passed as argument (is any). `--reset-values` and `--reuse-values` flags take precedence over `--reset-then-reuse-values`, making it ignored if one or the other is also used.

This is a slight improvement over the many times suggested workaround of `helm get values > v.yaml; helm upgrade --reset-values --values v.yaml` in the sense that this flag allows for atomic operation.

Closes #8085
Closes #3957

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/action/upgrade_test.go
Comment (jglick) [inline pkg/action/upgrade_test.go:297]: Maybe I am missing something, but where does this come from?

Comment (Okhoshi) [reply pkg/action/upgrade_test.go:297]: This is testing that values added to the chart between initial setup and upgrade are applied in the release values.

Comment (jglick) [reply pkg/action/upgrade_test.go:297]: Yeah I just could not find another occurrence of `256m` in this repository.

Comment (Okhoshi) [reply pkg/action/upgrade_test.go:297]: Is it an issue ?
At least we know for sure that the value is not coming from elsewhere 😅

Comment (jglick) [reply pkg/action/upgrade_test.go:297]: Finally got it—this variable is used twice. Sorry for noise.

**Comment by erickpeirson** [comment]
@bacongobbler How can we move this forward?

**Comment by gowtham-sai** [comment]
@bacongobbler Can you please shed some light here?

**Comment by bitjson** [comment]
This would really improve our workflow 👍 

Tested, and PR works as expected. Is there any additional work or review which needs to be done before this can be merged?

**Comment by zombrie** [comment]
This is exactly what we need for our current chart situation... is this going to be merged soon?

**Comment by yaron2** [comment]
Please merge this. I can't stress enough how broken the current experience is.

**Comment by Ga13Ou** [comment]
Any idea why this simple, yet useful pull request isn't merged yet?

**Comment by voidtek** [comment]
This seems can solve my actual headache. Any news?

**Comment by kostya2011** [comment]
Is there any news on this one?

**Comment by yaron2** [comment]
Looks like the project isn't maintained anymore.

**Comment by asauber** [comment]
This would be a huge help for getting the expected upgrade behavior when using `--set`. The PR also looks relatively easy to review. Can we merge this?

**Review by asauber — APPROVED** [review]
LGTM

**Review by michi-covalent — APPROVED** [review]

**Comment by mamoit** [comment]
@Okhoshi is there anything blocking this PR?

**Comment by mamoit** [comment]
Then @jglick, @asauber or @michi-covalent maybe?
Since they are the assigned reviewers.

**Comment by michi-covalent** [comment]
i don't have write access to the repo either. i just approved it because i love this PR ❤️

**Review by joejulian — APPROVED** [review]
I tested this and it works with every scenario I could think of. Approving but this still needs to pass CI.

**Comment by amannijhawan** [comment]
This is a quite useful PR, I have no special access either.

**Comment by 0x2b3bfa0** [comment]
@joejulian, it alreday passes continuous integration checks.

**Comment by devnicolas1** [comment]
Are there any plans to include this feature on 3.13?

**Review by scottrigby — APPROVED** [review]
LGTM. Thanks @Okhoshi for this contribution 👏 

A few extra kudos: the linked issues and PR description fully explain the use case for this new feature and the reasoning for this particular solution over existing workarounds. The PR includes automated tests, which are sensible and pass CI automation (since @joejulian last merged in main). This has been peer reviewed by other Helm community members, and reviewed now by two maintainers. Merging now, so this should be included in the next scheduled feature release. 🚀

**Comment by jglick** [comment]
> should be included in the next scheduled feature release

Per https://github.com/helm/helm/releases/tag/v3.13.2

> 3.14.0 is the next feature release and [will] be on January 17, 2024.

Looking forward to finally having this fixed!

**Comment by voidtek** [comment]
Can't wait for it to be merged into the release-3.13 branch.
https://github.com/helm/helm/commits/release-3.13

@scottrigby : any blockers?

**Comment by mikael-floden** [comment]
OMG finally! Can't believe this is true, I have spent hours and hours explaining how helm is broken to every new developer. How --reuse-values breaks the system, ignoring all new values from the next version/release 

People just didn't get it. I don't want to know how many weeks I have spent on users not understanding when you can and when you can't use --reuse-values.

This is how EVERYONE intuitively thinks it works, this parameter should be the default. The lack of this feature has hold back helm for to many years.

This is the "it always works parameter". Thanks you 🙏

**Pull request merged**
